cpp/coefficients.cpp

```diff
@@ -316,6 +306,229 @@ dolfinx_contact::pack_coefficient_quadrature(
+    std::shared_ptr<const dolfinx::fem::Function<PetscScalar>> coeff,
+    const int q_degree, std::span<const std::int32_t> active_entities,
+    dolfinx::fem::IntegralType integral)
+{
+
+  // Get mesh
+  std::shared_ptr<const dolfinx::mesh::Mesh> mesh
+      = coeff->function_space()->mesh();
+  assert(mesh);
+
+  // Get topology data
+  const dolfinx::mesh::Topology& topology = mesh->topology();
+  const std::size_t tdim = topology.dim();
+  const dolfinx::mesh::CellType cell_type = topology.cell_type();
+
+  // Get what entity type we are integrating over
+  std::size_t entity_dim;
+  switch (integral)
+  {
+  case dolfinx::fem::IntegralType::cell:
+    entity_dim = tdim;
+    break;
+  case dolfinx::fem::IntegralType::exterior_facet:
+    entity_dim = tdim - 1;
+    break;
+  default:
+    throw std::invalid_argument("Unsupported integral type.");
+  }
+
+  // Create quadrature rule
+  QuadratureRule q_rule(cell_type, q_degree, (int)entity_dim);
+
+  // Get element information
+  const dolfinx::fem::FiniteElement* element
+      = coeff->function_space()->element().get();
+  const std::size_t bs = element->block_size();
+  const std::size_t value_size = element->value_size();
+
+  // Tabulate function at quadrature points (assuming one derivatives)
+  dolfinx_contact::error::check_cell_type(cell_type);
+  const std::vector<double>& q_points = q_rule.points();
+  const std::vector<std::size_t>& q_offsets = q_rule.offset();
+  const std::size_t sum_q_points = q_offsets.back();
+  const std::size_t num_points_per_entity = q_offsets[1] - q_offsets[0];
+  std::array<std::size_t, 2> p_shape = {sum_q_points, tdim};
+  assert(q_rule.tdim() == tdim);
+
+  const basix::FiniteElement& basix_element = element->basix_element();
+  std::array<std::size_t, 4> tab_shape
+      = basix_element.tabulate_shape(1, sum_q_points);
+  std::vector<double> reference_basisb(
+      std::reduce(tab_shape.cbegin(), tab_shape.cend(), 1, std::multiplies{}));
+  element->tabulate(reference_basisb, q_points, p_shape, 1);
+  cmdspan4_t reference_basis(reference_basisb.data(), tab_shape);
+  assert(value_size / bs == tab_shape[3]);
+
+  // Get geometry data
+  const dolfinx::mesh::Geometry& geometry = mesh->geometry();
+  const std::size_t gdim = geometry.dim();
+  const dolfinx::graph::AdjacencyList<std::int32_t>& x_dofmap
+      = mesh->geometry().dofmap();
+  const dolfinx::fem::CoordinateElement& cmap = geometry.cmap();
+  const std::size_t num_dofs_g = cmap.dim();
+  std::span<const double> x_g = geometry.x();
+
+  // Tabulate coordinate basis to compute Jacobian
+  std::array<std::size_t, 4> c_shape = cmap.tabulate_shape(1, sum_q_points);
+  std::vector<double> c_basisb(
+      std::reduce(c_shape.cbegin(), c_shape.cend(), 1, std::multiplies{}));
+  cmap.tabulate(1, q_points, p_shape, c_basisb);
+  cmdspan4_t c_basis(c_basisb.data(), c_shape);
+
+  // Prepare geometry data structures
+  std::vector<double> Jb(gdim * tdim);
+  std::vector<double> Kb(tdim * gdim);
```

Comment:
Should we make these buffers `std::array<double, 9>` instead, as they require less memory. See: https://www.modernescpp.com/index.php/std-array-dynamic-memory-no-thanks